core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,449 @@
+     * the Unicode replacement character � (0xFFFD). Returning this value on failure is mandated
+     * by UTS-46.
+     */
+    def partiallyMappedInput: String
+
+    final override def getMessage: String =
+      toString
+
+    final override def toString: String =
+      s"MappingException(errors = ${errors}, partiallyMappedInput = ${partiallyMappedInput})"
+  }
+
+  object MappingException {
+    final private[this] case class MappingExceptionImpl(
+        override val errors: NonEmptyList[CodePointMappingException],
+        override val partiallyMappedInput: String)
+        extends MappingException
+
+    def apply(
+        errors: NonEmptyList[CodePointMappingException],
+        partiallyMappedInput: String): MappingException =
+      MappingExceptionImpl(errors, partiallyMappedInput)
+
+    implicit val hashAndOrderForMappingException
+        : Hash[MappingException] with Order[MappingException] =
+      new Hash[MappingException] with Order[MappingException] {
+        override def hash(x: MappingException): Int =
+          x.hashCode
+
+        override def compare(x: MappingException, y: MappingException): Int =
+          x.errors.compare(y.errors) match {
+            case 0 =>
+              x.partiallyMappedInput.compare(y.partiallyMappedInput)
+            case otherwise => otherwise
+          }
+      }
+  }
+
+  /**
+   * A constant for the Unicode replacement character �.
+   */
+  final private val ReplacementCharacter =
+    0xfffd
+}
+
+/**
+ * This is a base type which exists for the generated code.
+ */
+abstract private[uts46] class CodePointMapperBase {
+  protected def validAlways: BitSet
+
+  protected def validNV8: BitSet
+
+  protected def validXV8: BitSet
+
+  protected def ignored: BitSet
+
+  protected def mappedMultiCodePoints: IntMap[NonEmptyList[Int]]
+
+  protected def deviationMapped: IntMap[Int]
+
+  protected def deviationMultiMapped: IntMap[NonEmptyList[Int]]
+
+  protected def disallowed: BitSet
+
+  protected def disallowedSTD3Valid: BitSet
+
+  protected def disallowedSTD3Mapped: IntMap[Int]
+
+  protected def disallowedSTD3MultiMapped: IntMap[NonEmptyList[Int]]
+
+  protected def deviationIgnored: BitSet
+
+  protected def mapped: IntMap[Int]
+
+  /**
+   * The Unicode version the code point mapping functions target.
+   *
+   * @note
+   *   UTS-46 is backwards compatible with future versions of Unicode. The only part of the
+   *   mapping operation which can change between Unicode revisions is the disallowed character
+   *   set. This includes the STD3 Valid and STD3 mapped variants. In other words, code points
+   *   which were disallowed by some prior Unicode version may be made allowed in some future
+   *   Unicode version or they may still be disallowed but have their mappings under STD3
+   *   changed or they may be considered ''valid under STD3 only'', when they were formerlly
+   *   disllowed under STD3.
```

Comment:
Sorry I missed this before! But while we're here

```suggestion
   *   changed or they may be considered ''valid under STD3 only'', when they were formerly
   *   disallowed under STD3.
```